src/representations/representation_manager_spherical_invariants.hh

```diff
@@ -50,77 +50,180 @@
+    struct SphericalInvariantsPrecomputation<
+        SphericalInvariantsType::PowerSpectrum>
+        : SphericalInvariantsPrecomputationBase {
+      using Hypers_t = typename SphericalInvariantsPrecomputationBase::Hypers_t;
+      explicit SphericalInvariantsPrecomputation(const Hypers_t & hypers) {
         this->max_angular = hypers.at("max_angular");
+        // this->l_factors.resize(math::pow(this->max_angular + 1, 2_z));
+
+        // size_t lm{0};
+        // for (size_t l{0}; l < this->max_angular + 1; ++l) {
+        //   double l_factor{math::pow(std::sqrt(2 * l + 1), -1)};
+        //   for (size_t m{0}; m < 2 * l + 1; ++m) {
+        //     this->l_factors(lm) = l_factor;
+        //     ++lm;
+        //   }
+        // }
         this->l_factors.resize(this->max_angular + 1);
-
         for (size_t l{0}; l < this->max_angular + 1; ++l) {
-          double l_factor{math::pow(std::sqrt(2 * l + 1), -1)};
-          this->l_factors(l) = l_factor;
+          this->l_factors(l) = math::pow(std::sqrt(2 * l + 1), -1);
         }
       }
 
       size_t max_angular{0};
       //! factor of 1 / sqrt(2*l+1) in front of the powerspectrum
       Eigen::VectorXd l_factors{};
     };
+
+    template <>
+    struct SphericalInvariantsPrecomputation<
+        SphericalInvariantsType::BiSpectrum>
+        : SphericalInvariantsPrecomputationBase {
+      using Hypers_t = typename SphericalInvariantsPrecomputationBase::Hypers_t;
+      explicit SphericalInvariantsPrecomputation(const Hypers_t & hypers) {
+        this->max_angular = hypers.at("max_angular").get<size_t>();
+        this->inversion_symmetry = hypers.at("inversion_symmetry").get<bool>();
+        // get the number of non zero elements in the w3j
+        int n_elements{0};
+        for (size_t l1{0}; l1 < this->max_angular + 1; ++l1) {
+          for (size_t l2{0}; l2 < this->max_angular + 1; ++l2) {
+            for (size_t l3{0}; l3 < this->max_angular + 1; ++l3) {
+              if (l1 < static_cast<size_t>(std::abs<int>(l2 - l3)) ||
+                  l1 > l2 + l3) {
+                continue;
+              }
+              if (this->inversion_symmetry == true) {
+                if ((l1 + l2 + l3) % 2 == 1) {
+                  continue;
+                }
+              }
+              for (size_t m1{0}; m1 < 2 * l1 + 1; m1++) {
+                int m1s{static_cast<int>(m1 - l1)};
+                for (size_t m2{0}; m2 < 2 * l2 + 1; m2++) {
+                  int m2s{static_cast<int>(m2 - l2)};
+                  for (size_t m3{0}; m3 < 2 * l3 + 1; m3++) {
+                    int m3s{static_cast<int>(m3 - l3)};
+                    if (m1s + m2s + m3s != 0) {
+                      continue;
+                    }
+                    ++n_elements;
+                  }
+                }
+              }
+            }
+          }
+        }
+
+        this->w3js.resize(n_elements);
+        n_elements = 0;
+        wig_table_init(2 * (this->max_angular + 1), 3);
```

Comment:
And this isn't a big table where drag queens keep their wigs? Am I the only one who first thought of that? No...?

Moving on...

Reply:
you are too playful, I hadn't thought about that.

Reply:
Hehe, I thought about english royalty with their powdered wigs.

Reply:
Not too far off, when you think about it